frame/dex/src/lib.rs

```diff
@@ -0,0 +1,855 @@
+				}
+			}
+
+			amounts.reverse();
+			Ok(amounts)
+		}
+
+		fn get_amounts_out(
+			amount_in: &AssetBalanceOf<T>,
+			path: &BoundedVec<MultiAssetId<T::AssetId>, T::MaxSwapPathLength>,
+		) -> Result<Vec<AssetBalanceOf<T>>, DispatchError> {
+			let mut amounts: Vec<AssetBalanceOf<T>> = vec![*amount_in];
+
+			for assets_pair in path.windows(2) {
+				if let &[asset1, asset2] = assets_pair {
+					let (reserve_in, reserve_out) = Self::get_reserves(asset1, asset2)?;
+					let prev_amount = amounts.last().expect("Always has at least one element");
+					let amount_out = Self::get_amount_out(prev_amount, &reserve_in, &reserve_out)?;
+					amounts.push(amount_out);
+				}
+			}
+
+			Ok(amounts)
+		}
+
+		/// Used by the RPC service to provide current prices.
+		pub fn quote_price(
+			asset1: Option<u32>,
+			asset2: Option<u32>,
+			amount: u64,
+		) -> Option<AssetBalanceOf<T>> {
+			let into_multi = |asset| {
+				if let Some(asset) = asset {
+					MultiAssetId::Asset(T::AssetId::from(asset))
+				} else {
+					MultiAssetId::Native
+				}
+			};
+			let asset1 = into_multi(asset1);
+			let asset2 = into_multi(asset2);
+			let amount = amount.into();
+			let pool_id = Self::get_pool_id(asset1, asset2);
+			let pool_account = Self::get_pool_account(pool_id);
+			let (pool_asset1, _) = pool_id;
+
+			let balance1 = Self::get_balance(&pool_account, asset1);
+			let balance2 = Self::get_balance(&pool_account, asset2);
+			if !balance1.is_zero() {
+				let (reserve1, reserve2) =
+					if asset1 == pool_asset1 { (balance1, balance2) } else { (balance2, balance1) };
+				Self::quote(&amount, &reserve1, &reserve2).ok()
+			} else {
+				None
+			}
+		}
+
+		/// Calculates the optimal amount from the reserves.
+		pub fn quote(
+			amount: &AssetBalanceOf<T>,
+			reserve1: &AssetBalanceOf<T>,
+			reserve2: &AssetBalanceOf<T>,
+		) -> Result<AssetBalanceOf<T>, Error<T>> {
+			// amount * reserve2 / reserve1
+			Self::mul_div(amount, reserve2, reserve1)
+		}
+
+		fn calc_lp_amount_for_zero_supply(
+			amount1: &AssetBalanceOf<T>,
+			amount2: &AssetBalanceOf<T>,
+		) -> Result<AssetBalanceOf<T>, Error<T>> {
+			let amount1 =
+				u128::try_from(*amount1).map_err(|_| Error::<T>::UnsupportedBalanceType)?;
+			let amount2 =
+				u128::try_from(*amount2).map_err(|_| Error::<T>::UnsupportedBalanceType)?;
+
+			let result = amount1
+				.checked_mul(amount2)
+				.ok_or(Error::<T>::Overflow)?
+				.integer_sqrt()
+				.checked_sub(MIN_LIQUIDITY.into())
+				.ok_or(Error::<T>::Overflow)?;
+
+			result.try_into().map_err(|_| Error::<T>::Overflow)
+		}
+
+		fn mul_div(
+			a: &AssetBalanceOf<T>,
+			b: &AssetBalanceOf<T>,
+			c: &AssetBalanceOf<T>,
+		) -> Result<AssetBalanceOf<T>, Error<T>> {
+			let a = u128::try_from(*a).map_err(|_| Error::<T>::UnsupportedBalanceType)?;
+			let b = u128::try_from(*b).map_err(|_| Error::<T>::UnsupportedBalanceType)?;
+			let c = u128::try_from(*c).map_err(|_| Error::<T>::UnsupportedBalanceType)?;
+
+			let result = a
```

Comment:
this need to promote to U256 to avoid overflow, which is going to happen

Reply:
It's promoting from u64 already. U256 might be overkill, but types are plugable so maybe we need a trait to be able to promote into a larger numeric type? (and a trait to try and get back)

Reply:
you have to support u128 balance type